qiime2/metadata.py

```diff
@@ -14,6 +14,38 @@
 
 class Metadata:
     def __init__(self, dataframe):
+        # `/` and `\0` aren't permitted because they are invalid filename
+        # characters on *nix filesystems. The remaining values aren't permitted
+        # because they *could* be misinterpreted by a shell (e.g. `*`, `|`).
+        illegal_chars = ['/', '\0', '\\', '*', '<', '>', '?', '|', '$']
+        illegal_chars = set(illegal_chars)
+        chars_for_msg = ", ".join("%r" % i for i in illegal_chars)
+
+        categories = dataframe.columns.values.tolist()
+        indices = dataframe.index.tolist()
+
+        if dataframe.columns.dtype_str != 'object':
+            msg = "Non-string Metadata category labels detected"
+            raise ValueError(invalid_metadata_template % msg)
+
+        if any(illegal_chars & set(category) for category in categories):
+            msg = "Invalid characters (e.g. %s) detected in metadata " \
+                  "category labels" % chars_for_msg
```

Comment:
I'm ok with this (and the conditional is very pretty), but it may be worth going to the extra mile to figure out what columns are wrong and why.